packages/devtools_app/lib/src/screens/memory/framework/offline_data/offline_data.dart

```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../../panes/chart/controller/chart_pane_controller.dart';
+import '../../panes/diff/controller/diff_pane_controller.dart';
+import '../../panes/profile/profile_pane_controller.dart';
+import '../../shared/heap/class_filter.dart';
+
+class _Json {
+  static const selectedTab = 'selectedTab';
+  static const classFilter = 'classFilter';
+  static const diffData = 'diffData';
+  static const profileData = 'profileData';
+  static const chartData = 'chartData';
+}
+
+class OfflineMemoryData {
+  OfflineMemoryData(
+    this.diff,
+    this.profile,
+    this.chart,
+    this.filter, {
+    this.isEmpty = false,
+    required this.selectedTab,
+  });
+
+  factory OfflineMemoryData.parse(Map<String, dynamic> json) {
+    Map<String, dynamic> item(String key) =>
+        json[key] as Map<String, dynamic>? ?? {};
+    return OfflineMemoryData(
+      DiffPaneController.parse(item(_Json.diffData)),
+      ProfilePaneController.parse(item(_Json.profileData)),
+      MemoryChartPaneController.parse(item(_Json.chartData)),
+      ClassFilter.parse(item(_Json.classFilter)),
+      selectedTab: json[_Json.selectedTab] as int? ?? 0,
+      isEmpty: json.isEmpty,
+    );
+  }
+
+  final bool isEmpty;
+
+  final int selectedTab;
+  final ClassFilter filter; // filter is shared between tabs, so it's here
+
+  final DiffPaneController diff;
+  final ProfilePaneController profile;
+  final MemoryChartPaneController chart;
+
+  Map<String, dynamic> prepareForOffline() {
```

Comment:
It is consistent with OfflineMemoryData. 

I would update both with what is recommended for serialization: https://docs.flutter.dev/data-and-backend/serialization/json#serializing-json-inside-model-classes

How does it sound?

Reply:
If is recommended to never use `dynamic`. We should be using `Object?` instead. These have different implications in the compiler and for type checking. Couldn't find official docs but here's an issue with some discussion on this topic: https://github.com/dart-lang/language/issues/3192

Reply:
I suggest to do it as separate clean up, including `OfflineMemoryData.parse(Map<String, dynamic> json)` and others.
I expect unexpected side effects that may cause us to change our mind here. I remember trying to use Object for json instead of dynamic and facing some difficulties that convinced me to stay with standard for json.

Serialization is a specific mechanism that picks up all objects that meet this standard:  https://docs.flutter.dev/data-and-backend/serialization/json#serializing-json-inside-model-classes

This mechanism may break with switch to Object.

Reply:
@srawlins do you know if the linked docs above are up to date? I know you've done a pretty large overhaul of all the JSON parsing we were doing in DevTools using extension types. Can you advise here WRT using `dynamic` vs `Object?`?

Reply:
I think the docs are still accurate. Anything produced by `dart:convert` uses `dynamic` (e.g. `List<dynamic>` and `Map<String, dynamic>` or maybe even `Map<dynamic, dynamic>`) instead of `Object?`.

We can also reference the [general recommendation to avoid `dynamic`](https://dart.dev/effective-dart/design#avoid-using-dynamic-unless-you-want-to-disable-static-checking), and I stick strongly to that. I avoid `dynamic` anywhere it isn't necessary. So then then question is: when working with objects produced by `dart:convert`, what should we use? I think `Object?` is safe everywhere here, but I actually haven't worked through the edge cases, etc. I can't promise it, so I think I used `dynamic` in a lot of my devtools PRs, just for consistency.

The runtime danger is that the objects returned from `dart:convert` don't "promote" types of objects when you might expect it. For example, if you call `x = jsonDecode('["hello", "there"]');`, you get a `List<dynamic>` thing at runtime, and you might think "well, I know it's always a List of Strings, so I'll use `x as List<String>`." That is ok at compile-time, but will throw at runtime. You might try `if (x is List<String>)` as defensive, but now that condition _will never be true_ at runtime. It's super awkward but you have to do `if (x is List) x.cast<String>()` (which could still throw if you have bad data) or (`if (x is List) for (var y in x) if (y is String) ...`). Super tedious.

Back to `Object?` vs `dynamic`. Either of these will work, I think all the time, but I'm not sure. E.g.

```dart
import 'dart:convert';

void main() {
  var b = jsonDecode('[1, "hello"]');
  if (b is List<dynamic>) {
    print('yay a List<dynamic>');
  }
  if (b is List<Object?>) {
    print('yay a List<Object?>');
  }
  if (b is List<Object>) {
    print('yay a List');
  }
  f1();
  f2();
  f3();
}

List<dynamic> f1() => jsonDecode('[1, "hello"]');
List<Object?> f2() => jsonDecode('[1, "hello"]');
List<Object> f3() => jsonDecode('[1, "hello"]');
```

This prints "yay a List<dynamic>" and "yay a List<Object?>" because of how is-checks work (assignability?). But not "yay a List<Object>". That test fails. And the calls to `f1` and `f2` pass, but the call to `f3` throws an exception, because the returned object is not a `List<Object>`.

I think guidance w.r.t. dynamic-and-JSON is a team-by-team decision. We can't all become type-system experts and JSON experts (like I can never if JSON explicitly _does not_ support a `null` concept, or does, or if Dart does something special...). So whatever consistent policy keeps the team writing safe code and understandable (for the team) code, is a good policy. I don't think analyzer team has a consistent policy, and we definitely could have JSON-parsing bugs, or YAML-parsing bugs, in malformed analysis_options.yaml, or package_config.json, etc. It's something I keep punting on, hoping `dart:convert` will just change to returning `Object?` objects instead of `dynamic`, or maybe a better JSON-parsing story. 🤷

Reply:
Thanks. I suggest to stay compliant with public recommendation.
If we want to invest into switch to <String, Object?>, we need to reflect it in our style rules and migrate the entire project at once, to keep it consistent.